enhancements/etcd/protecting-etcd-quorum-during-control-plane-scaling.md

```diff
@@ -0,0 +1,450 @@
+  and MachineSets from the manifests directory.
+  - This was a common scenario when vSphere was introduced as a new Machine API provider. The instruction was missed
+    from the UPI documentation
+  - In this scenario, the Machines for the Control Plane all siti in the `Provisioning` phase
+  - The Machine objects created do not have a full configuration and as such, the Machine API fails to provision new
+    Control Plane Machines
+- The cluster is created using either the UPI or IPI installation method, but to an unsupported platform (eg. platform
+  None)
+  - In this case, as Machine API does not support the platform, the installer will not have generated and Machines/
+    MachineSets
+
+In each of the scenarios, either that are no Machines in the cluster or the Machines will be stuck in the
+`Provisioning` phase. In particular the absence of any `Running` Machine signals that the Machine API is non-functional.
+
+##### New metrics to export via the telemeter
+
+TODO: Consult with SREs to identify metrics they might find useful for us to export
+
+#### Example of vertically scaling a Control Plane Machine
+
+To use this mechanism to vertically scale a Control Plane, the following procedure must be carried out by some user or
+operator.
+
+1. Identify a Controle Plane Machine for replacement (eg. `my-cluster-master-0`)
+1. Determine a new name for the new Control Plane Machine that will not conflict with existing Control Plane Machines
+   (eg `my-cluster-master-3`)
+1. Take a copy of the Machine resource: `oc get machine -n openshift-machine-api my-cluster-master-0 -o yaml > my-
+   cluster-master-3.yaml`
+1. Modify the Machine YAML to update the `name` and the size of the instance (eg. changing
+   `spec.providerSpec.value.instanceType` from `c5.xlarge` to `c5.2xlarge`)
+1. Create the new Machine: `oc create -f my-cluster-master-3.yaml`
+1. Delete the old Machine: `oc delete machine -n openshift-machine-api my-cluster-master-0`
+1. Wait until the process described in [Operational flow during a resize operation](#Operational-flow-during-a-resize-operation)
+   has been completed. The end result of this is that the Machine `my-cluster-master-0` will be removed
+
+### Risks and Mitigations
+
+#### Blocking removal of Machines in environments with restricted capacity
+
+This proposal assumes that when a Machine is to be replaced, that there is additional capacity available to allow the
+new Machine to be created before the old Machine is removed. This may not be true in all environments, for example if
+there is no quota left or in bare-metal environments with limited hardware available.
+
+To ensure that we do not block users from replacing Control Plane Machines in these scenarios, we must allow a user to
+remove the etcd quorum hook without it being replaced by the etcd Operator.
+When a Machine is marked for deletion, if the hook is removed, the etcd Operator will not replace it. This will allow
```

Comment:
> the etcd Operator will not replace it

what does this mean? It will not readd the hook? So there must be a check in the hook-add controller that it only adds hooks to unmarked machines. This must be added somewhere above, at least as a note.

Reply:
Is "mark for deletion" the same as setting the `metadata.deleteTimestamp`? is it possible to mutate a deleted, but not removed object?

Reply:
> what does this mean? It will not readd the hook? So there must be a check in the hook-add controller that it only adds hooks to unmarked machines.

Yep, that's what we are trying to say here, will see if there's an appropriate place to add a note above. This is called out in more detail in the machine deletion hooks enhancement though, that's why I didn't go into super detail level here

> Is "mark for deletion" the same as setting the metadata.deleteTimestamp? is it possible to mutate a deleted, but not removed object?

Yes, I mean someone has called a DELETE operation on the resource so now the metadata.deletionTimestamp is set, not sure if there's a more commonly used verbiage for this than "marked for deletion"?

It is possible to mutate an object that has not been deleted, the API server doesn't prevent this. We can mutate it all the way up until the GC removes it, but that will only happen when the Machine API removes its finaliser. I've tested this manually and the deletion hooks have gone through QE.

Reply:
sounds good, just clarify here a bit.